File: frontend/src/utils/coordinateMapper.ts
import type { CSSProperties } from 'react'
import type { Bounds } from '../types'

/** Single coordinate system definition shared by overlays and hit-testing. */
export interface CoordinateMapping {
  /** XML / UIAutomator bounds space */
  hierarchyWidth: number
  hierarchyHeight: number
  /** PNG pixel space (must match loaded image naturalWidth/Height) */
  screenshotWidth: number
  screenshotHeight: number
  /** Optional origin offset when hierarchy root != screenshot origin */
  offsetX: number
  offsetY: number
  scaleX: number
  scaleY: number
}

export interface Point {
  x: number
  y: number
}

export interface DisplayRect {
  left: number
  top: number
  width: number
  height: number
}

export interface HitTestDebug {
  rawClient: Point
  displayRect: DisplayRect
  containerScroll: Point
  panOffset: Point
  zoom: number
  screenshotPixel: Point | null
  hierarchyPixel: Point | null
  mapping: CoordinateMapping
}

export function buildMapping(
  hierarchyWidth: number,
  hierarchyHeight: number,
  screenshotWidth: number,
  screenshotHeight: number,
  offsetX = 0,
  offsetY = 0,
): CoordinateMapping {
  const hw = hierarchyWidth || 1
  const hh = hierarchyHeight || 1
  const sw = screenshotWidth || hw
  const sh = screenshotHeight || hh
  return {
    hierarchyWidth: hw,
    hierarchyHeight: hh,
    screenshotWidth: sw,
    screenshotHeight: sh,
    offsetX,
    offsetY,
    scaleX: hw / sw,
    scaleY: hh / sh,
  }
}

/**
 * Map viewport mouse position to screenshot PNG pixel coordinates.
 * Uses the rendered image rect (includes pan/zoom transforms) and native PNG dimensions.
 */
export function displayToScreenshot(
  clientX: number,
  clientY: number,
  imageRect: DisplayRect,
  nativeWidth: number,
  nativeHeight: number,
): Point | null {
  if (!imageRect.width || !imageRect.height || !nativeWidth || !nativeHeight) {
    return null
  }
  const relX = clientX - imageRect.left
  const relY = clientY - imageRect.top
  if (relX < 0 || relY < 0 || relX > imageRect.width || relY > imageRect.height) {
    return null
  }
  return {
    x: Math.round((relX / imageRect.width) * nativeWidth),
    y: Math.round((relY / imageRect.height) * nativeHeight),
  }
}

export function screenshotToHierarchy(
  x: number,
  y: number,
  mapping: CoordinateMapping,
): Point {
  const { hierarchyWidth, hierarchyHeight, screenshotWidth, screenshotHeight, offsetX, offsetY } = mapping
  if (!screenshotWidth || !screenshotHeight) {
    return { x: Math.round(x), y: Math.round(y) }
  }
  const hx = Math.round(x * (hierarchyWidth / screenshotWidth) + offsetX)
  const hy = Math.round(y * (hierarchyHeight / screenshotHeight) + offsetY)
  return {
    x: Math.max(0, Math.min(hx, Math.max(hierarchyWidth - 1, 0))),
    y: Math.max(0, Math.min(hy, Math.max(hierarchyHeight - 1, 0))),
  }
}

export function hierarchyToScreenshot(bounds: Bounds, mapping: CoordinateMapping): Bounds {
  const sx = mapping.screenshotWidth / mapping.hierarchyWidth
  const sy = mapping.screenshotHeight / mapping.hierarchyHeight
  return {
    x1: Math.round((bounds.x1 - mapping.offsetX) * sx),
    y1: Math.round((bounds.y1 - mapping.offsetY) * sy),
    x2: Math.round((bounds.x2 - mapping.offsetX) * sx),
    y2: Math.round((bounds.y2 - mapping.offsetY) * sy),
  }
}

/** CSS overlay percentages on a box sized to native screenshot dimensions. */
export function boundsToOverlayStyle(bounds: Bounds, mapping: CoordinateMapping): CSSProperties {
  const ss = hierarchyToScreenshot(bounds, mapping)
  const sw = mapping.screenshotWidth || 1
  const sh = mapping.screenshotHeight || 1
  return {
    left: `${(ss.x1 / sw) * 100}%`,
    top: `${(ss.y1 / sh) * 100}%`,
    width: `${((ss.x2 - ss.x1) / sw) * 100}%`,
    height: `${((ss.y2 - ss.y1) / sh) * 100}%`,
  }
}

export function boundsCenter(bounds: Bounds): Point {
  return {
    x: Math.round((bounds.x1 + bounds.x2) / 2),
    y: Math.round((bounds.y1 + bounds.y2) / 2),
  }
}

export function buildHitTestDebug(
  clientX: number,
  clientY: number,
  imageRect: DisplayRect,
  nativeWidth: number,
  nativeHeight: number,
  mapping: CoordinateMapping,
  containerScroll: Point,
  panOffset: Point,
  zoom: number,
): HitTestDebug {
  const screenshotPixel = displayToScreenshot(clientX, clientY, imageRect, nativeWidth, nativeHeight)
  const hierarchyPixel = screenshotPixel ? screenshotToHierarchy(screenshotPixel.x, screenshotPixel.y, mapping) : null
  return {
    rawClient: { x: clientX, y: clientY },
    displayRect: imageRect,
    containerScroll,
    panOffset,
    zoom,
    screenshotPixel,
    hierarchyPixel,
    mapping,
  }
}

File: backend/inspectiq/engine/coordinate_mapper.py
"""Map between screenshot pixel coordinates and UI hierarchy bounds space."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

from inspectiq.domain.models import Bounds, ElementNode


@dataclass
class CoordinateMapping:
    """Authoritative mapping between PNG screenshot pixels and XML hierarchy bounds."""

    device_width: int
    device_height: int
    hierarchy_width: int
    hierarchy_height: int
    screenshot_width: int
    screenshot_height: int
    scale_x: float
    scale_y: float
    offset_x: int = 0
    offset_y: int = 0
    rotation: int = 0

    def to_dict(self) -> dict:
        return {
            "device_width": self.device_width,
            "device_height": self.device_height,
            "hierarchy_width": self.hierarchy_width,
            "hierarchy_height": self.hierarchy_height,
            "screenshot_width": self.screenshot_width,
            "screenshot_height": self.screenshot_height,
            "scale_x": round(self.scale_x, 6),
            "scale_y": round(self.scale_y, 6),
            "offset_x": self.offset_x,
            "offset_y": self.offset_y,
            "rotation": self.rotation,
        }


def hierarchy_dimensions(tree: ElementNode, fallback: Tuple[int, int] = (1080, 1920)) -> Tuple[int, int]:
    """Infer hierarchy width/height from parsed XML bounds (max x2/y2)."""
    max_x = 0
    max_y = 0

    def walk(node: ElementNode) -> None:
        nonlocal max_x, max_y
        if node.bounds:
            max_x = max(max_x, node.bounds.x2)
            max_y = max(max_y, node.bounds.y2)
        for child in node.children:
            walk(child)

    walk(tree)
    w = max_x or fallback[0]
    h = max_y or fallback[1]
    return w, h


def hierarchy_origin(tree: ElementNode) -> Tuple[int, int]:
    """Root bounds origin — informational only; UIAutomator uses absolute screen coords."""
    if tree.bounds:
        return tree.bounds.x1, tree.bounds.y1
    return 0, 0


def build_coordinate_mapping(
    tree: Optional[ElementNode],
    device_size: Tuple[int, int],
    screenshot_size: Tuple[int, int],
    rotation: int = 0,
) -> CoordinateMapping:
    device_w, device_h = device_size
    screenshot_w, screenshot_h = screenshot_size
    fallback = device_size if device_w and device_h else (1080, 1920)
    hierarchy_w, hierarchy_h = hierarchy_dimensions(tree, fallback) if tree else fallback
    if not screenshot_w or not screenshot_h:
        screenshot_w, screenshot_h = hierarchy_w, hierarchy_h
    scale_x = hierarchy_w / screenshot_w if screenshot_w else 1.0
    scale_y = hierarchy_h / screenshot_h if screenshot_h else 1.0
    return CoordinateMapping(
        device_width=device_w or hierarchy_w,
        device_height=device_h or hierarchy_h,
        hierarchy_width=hierarchy_w,
        hierarchy_height=hierarchy_h,
        screenshot_width=screenshot_w,
        screenshot_height=screenshot_h,
        scale_x=scale_x,
        scale_y=scale_y,
        offset_x=0,
        offset_y=0,
        rotation=rotation,
    )


def screenshot_to_hierarchy(
    x: float,
    y: float,
    hierarchy_w: int,
    hierarchy_h: int,
    screenshot_w: int,
    screenshot_h: int,
) -> Tuple[int, int]:
    """Convert screenshot pixel coordinates to hierarchy/XML coordinate space."""
    if screenshot_w <= 0 or screenshot_h <= 0:
        return int(round(x)), int(round(y))
    hx = int(round(x * hierarchy_w / screenshot_w))
    hy = int(round(y * hierarchy_h / screenshot_h))
    hx = max(0, min(hx, max(hierarchy_w - 1, 0)))
    hy = max(0, min(hy, max(hierarchy_h - 1, 0)))
    return hx, hy


def hierarchy_to_screenshot_bounds(
    bounds: Bounds,
    mapping: CoordinateMapping,
) -> Bounds:
    sx = mapping.screenshot_width / mapping.hierarchy_width if mapping.hierarchy_width else 1.0
    sy = mapping.screenshot_height / mapping.hierarchy_height if mapping.hierarchy_height else 1.0
    return Bounds(
        x1=int(round((bounds.x1 - mapping.offset_x) * sx)),
        y1=int(round((bounds.y1 - mapping.offset_y) * sy)),
        x2=int(round((bounds.x2 - mapping.offset_x) * sx)),
        y2=int(round((bounds.y2 - mapping.offset_y) * sy)),
    )


def hierarchy_to_screenshot_pct(
    bounds: Bounds,
    hierarchy_w: int,
    hierarchy_h: int,
) -> dict[str, str]:
    """CSS percentage overlay within a screenshot-sized box (non-uniform scale safe)."""
    if hierarchy_w <= 0 or hierarchy_h <= 0:
        hierarchy_w = hierarchy_w or 1
        hierarchy_h = hierarchy_h or 1
    return {
        "left": f"{(bounds.x1 / hierarchy_w) * 100}%",
        "top": f"{(bounds.y1 / hierarchy_h) * 100}%",
        "width": f"{((bounds.x2 - bounds.x1) / hierarchy_w) * 100}%",
        "height": f"{((bounds.y2 - bounds.y1) / hierarchy_h) * 100}%",
    }


def bounds_debug_info(bounds: Bounds) -> dict:
    return {
        "bounds": [bounds.x1, bounds.y1, bounds.x2, bounds.y2],
        "width": bounds.x2 - bounds.x1,
        "height": bounds.y2 - bounds.y1,
        "center_x": (bounds.x1 + bounds.x2) // 2,
        "center_y": (bounds.y1 + bounds.y2) // 2,
    }
